vnext/Microsoft.ReactNative.Cxx/PropertyBag.h

```diff
@@ -0,0 +1,250 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#ifndef MICROSOFT_REACTNATIVE_PROPERTYBAG
+#define MICROSOFT_REACTNATIVE_PROPERTYBAG
+
+//!
+//! PropertyBag is a thread-safe storage of property values.
+//! Internally the value is IInspectable and the key is a name object that represents an
+//! atomized string name. Each name object is defined in the context of a namespace object.
+//! The null or empty namespace object is a global namespace object.
+//! The property name is unique for the same namespace object.
+//! Different namespaces may have properties with the same local names.
+//!
+//! The PropertyBag struct is a wrapper around the ABI-safe IReactPropertyBag interface.
+//! The IReactPropertyBag represents all values as IInspectable object which can wrap any type.
+//! On top of the untyped IReactPropertyBag values, the PropertyBag offers a set of typed property accessors:
+//! Get, GetOrCreate, Remove, and Set.
+//!
+//! To simplify access to properties we offer PropertyName type that helps to
+//! wrap up property name and type into one variable.
+//!
+//! For example, we can define a property to access an integer value:
+//!
+//!     const React::PropertyName<int> MyIntProperty{"MyInt"};
```

Comment:
>int [](start = 34, length = 3)

TMP newb speaking, not sure if this can work, but could, for better readability, the property name become a template argument, e.g.

React::Property<int, "MyInt"> myInt;

Reply:
It is a good question! It is one of the most popular questions in C++ forums. I probably spent a week looking for various work-arounds. No, it is not possible to do in a simple way. It could become possible in C++23. But, I agree, it would be nice.

Reply:
I *think* you can do it if you have a global/static const char* with the string and use that, since that pointer gets treated as an int by the template but yes it isn't super clean

Reply:
Yes, you are right. Use of static strings is one of the ways we can use strings in templates, but it requires to use a variable, not a string literal, in the template parameters, and like you said: "it isn't super clean".